packages/discord.js/src/structures/UncachedGuildMember.js

```diff
@@ -0,0 +1,328 @@
+     */
+    this.parsedPermissions = new PermissionsBitField(data.permissions).freeze();
+
+    /**
+     * Raw total permissions of the member in this channel, including overwrites
+     * @type {?string}
+     * @deprecated Use {@link UncachedGuildMember#parsedPermissions} instead.
+     * This field will be replaced with parsedPermissions in the future.
+     */
+    this.permissions = data.permissions;
+
+    /**
+     * The timestamp this member's timeout will be removed
+     * @type {?number}
+     */
+    this.communicationDisabledUntilTimestamp =
+      data.communication_disabled_until && Date.parse(data.communication_disabled_until);
+
+    /**
+     * The timestamp this member's timeout will be removed, as an ISO8601 timestamp
+     * @type {?string}
+     * @deprecated Use {@link UncachedGuildMember#communicationDisabledUntilTimestamp}
+     * or {@link UncachedGuildMember#communicationDisabledUntil} instead
+     */
+    this.communication_disabled_until = data.communication_disabled_until;
+
+    /**
+     * Whether this UncachedGuildMember is a partial (always true, as it is a partial GuildMember)
+     * @type {boolean}
+     * @readonly
+     */
+    this.partial = true;
+  }
+
+  /**
+   * A link to the member's guild avatar.
+   * @param {ImageURLOptions} [options={}] Options for the image URL
+   * @returns {?string}
+   */
+  avatarURL(options = {}) {
+    return this.avatar && this.client.rest.cdn.guildMemberAvatar(this.guildId, this.id, this.avatar, options);
+  }
+
+  /**
+   * A link to the member's guild avatar if they have one.
+   * Otherwise, a link to their {@link User#displayAvatarURL} will be returned.
+   * @param {ImageURLOptions} [options={}] Options for the Image URL
+   * @returns {string}
+   */
+  displayAvatarURL(options) {
+    return this.avatarURL(options) ?? this.user.displayAvatarURL(options);
+  }
+
+  /**
+   * The time this member joined the guild
+   * @type {?Date}
+   * @readonly
+   */
+  get joinedAt() {
+    return this.joinedTimestamp && new Date(this.joinedTimestamp);
+  }
+
+  /**
+   * The time this member's timeout will be removed
+   * @type {?Date}
+   * @readonly
+   */
+  get communicationDisabledUntil() {
+    return this.communicationDisabledUntilTimestamp && new Date(this.communicationDisabledUntilTimestamp);
+  }
+
+  /**
+   * The last time this member started boosting the guild
+   * @type {?Date}
+   * @readonly
+   */
+  get premiumSince() {
+    return this.premiumSinceTimestamp && new Date(this.premiumSinceTimestamp);
+  }
+
+  /**
+   * The member's id
+   * @type {Snowflake}
+   * @readonly
+   */
+  get id() {
+    return this.user.id;
+  }
+
+  /**
+   * The DM between the client's user and this member
+   * @type {?DMChannel}
+   * @readonly
+   */
+  get dmChannel() {
+    return this.client.users.dmChannel(this.id);
+  }
+
+  /**
+   * The nickname of this member, or their user display name if they don't have one
+   * @type {?string}
```

Comment:
would they not always have a _display_ name, or is this something to do with partials?

Reply:
hm yeah, why is this nullable in GuildMember? also user is nullable in GuildMember but id (which just returns this.user.id) isn't
are both of those just incorrectly marked as nullable

Reply:
it would seem so

Reply:
Is the user being nullable related to the
> The field `user` won't be included in the member object attached to `MESSAGE_CREATE` and `MESSAGE_UPDATE` gateway events.

note that is under the Guild Member object table [in the DAPI docs](https://discord.com/developers/docs/resources/guild#guild-member-object)?

Reply:
this sounds like something we need to investigate.... possibly seperately
but that would mean the docs on here could be wrong until investigation
complete

_re posted because email replies dont work properly_

Reply:
In those two events the `message.author` field contains the `user` and is patched into the `GuildMember` accordingly so won’t be `null` either https://github.com/discordjs/discord.js/blob/de14c92c1158d3e5d7d87d29d2fe9d99eb407df5/packages/discord.js/src/structures/Message.js#L339

Reply:
so in conclusion: the docs are wrong